File: scripts/day6-part1-optimized.py
import time

def solve_part1(input_data):
    lines = input_data.split('\n')
    width = len(lines[0])
    height = len(lines)
    
    grand_total = 0
    col = 0
    
    while col < width:
        # Skip empty columns
        has_non_space = False
        for row in range(height):
            if col < len(lines[row]) and lines[row][col] != ' ':
                has_non_space = True
                break
        
        if not has_non_space:
            col += 1
            continue
        
        # Find the end of this problem
        end_col = col
        while end_col < width:
            is_all_space = True
            for row in range(height):
                if end_col < len(lines[row]) and lines[row][end_col] != ' ':
                    is_all_space = False
                    break
            if is_all_space:
                break
            end_col += 1
        
        # Extract and calculate this problem
        result = 0
        operator = '+'
        
        for row in range(height):
            cell = lines[row][col:end_col].strip()
            if cell:
                if cell == '+':
                    operator = '+'
                elif cell == '*':
                    operator = '*'
                    result = result or 1
                else:
                    num = int(cell)
                    if operator == '+':
                        result += num
                    else:
                        result = (result or 1) * num
        
        grand_total += result
        col = end_col
    
    return grand_total

# Read input
with open("../inputs/day-6-input.txt", "r") as f:
    input_data = f.read().strip()

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part1(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 6 Part 1 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")

File: scripts/day6-part1-optimized.js
const fs = require("fs");

function solvePart1(input) {
  const lines = input.split("\n");
  const width = lines[0].length;
  const height = lines.length;

  let grandTotal = 0;
  let col = 0;

  while (col < width) {
    // Skip empty columns
    let hasNonSpace = false;
    for (let row = 0; row < height; row++) {
      if (lines[row][col] && lines[row][col] !== " ") {
        hasNonSpace = true;
        break;
      }
    }

    if (!hasNonSpace) {
      col++;
      continue;
    }

    // Find the end of this problem
    let endCol = col;
    while (endCol < width) {
      let isAllSpace = true;
      for (let row = 0; row < height; row++) {
        if (lines[row][endCol] && lines[row][endCol] !== " ") {
          isAllSpace = false;
          break;
        }
      }
      if (isAllSpace) break;
      endCol++;
    }

    // Extract this problem
    let result = 0;
    let operator = "+";

    for (let row = 0; row < height; row++) {
      const cell = lines[row].substring(col, endCol).trim();
      if (cell) {
        if (cell === "+") {
          operator = "+";
        } else if (cell === "*") {
          operator = "*";
          result = result || 1;
        } else {
          const num = parseInt(cell);
          if (operator === "+") {
            result += num;
          } else {
            result = (result || 1) * num;
          }
        }
      }
    }

    grandTotal += result;
    col = endCol;
  }

  return grandTotal;
}

const input = fs.readFileSync("../inputs/day-6-input.txt", "utf-8").trim();

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart1(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 6 Part 1 - Optimized)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
